src/openvpn/xkey_provider.c

```diff
@@ -88,13 +88,26 @@ typedef struct
     EVP_PKEY *pubkey;
     /* origin of key -- native or external */
     XKEY_ORIGIN origin;
+    /* sign function in backend to call */
+    XKEY_EXTERNAL_SIGN_fn *sign;
+    /* keydata handle free function of backend */
+    XKEY_PRIVKEY_FREE_fn *free;
     XKEY_PROVIDER_CTX *prov;
     int refcount;                /* reference count */
 } XKEY_KEYDATA;
 
 #define KEYTYPE(key) ((key)->pubkey ? EVP_PKEY_get_id((key)->pubkey) : 0)
 #define KEYSIZE(key) ((key)->pubkey ? EVP_PKEY_get_size((key)->pubkey) : 0)
```

Comment:
These are probably fine as they are. Otherwise maybe writing them as small static inline functions could also work if want a bit of extra type safety.

Reply:
Once pre-processed, these are just inline code and thus type safe, isn't it?